src/simian/util/simianfacter

```diff
@@ -54,21 +56,35 @@ def GetSerialNumber():
 
 def GetFacterFacts():
   """Returns a dictionary of facter facts."""
-  return_code, stdout, unused_stderr = Exec(['facter', '-p'])
+  facter_paths = ('/usr/bin/facter', '/opt/puppetlabs/bin/puppet')
+  facter_path = None
+  for path in facter_paths:
+    if os.path.exists(path):
+      facter_path = path
+      break
+
+  if facter_path == '/usr/bin/facter':
+    return_code, stdout, unused_stderr = Exec([facter_path, '-p'])
+  else:
+    return_code, stdout, unused_stderr = Exec([facter_path, 'facts', '--render-as', 'json'])
 
   # If execution of factor was successful build the client identifier
   if return_code != 0:
     return {}
 
   facts = {}
 
-  # Iterate over the facter output and create a dictionary of the output
-  lines = stdout.splitlines()
-  for line in lines:
-    (key, unused_sep, value) = line.split(' ', 2)
-    value = value.strip()
-    facts[key] = value
-  return facts
+  # Convert facter output to a dictionary
+  if facter_path == '/usr/bin/facter':
+    lines = stdout.splitlines()
+    for line in lines:
+      (key, unused_sep, value) = line.split(' ', 2)
+      value = value.strip()
+      facts[key] = value
+    return facts
+  else:
+    facts = json.loads(stdout)
```

Comment:
pardon, it's confusing because you're highlighting the code that handles the _previous_ version of facter, which is the same... I took another look just now, and any custom facts (as site and all of the ones Simian cares about are flat, un='structured' facts,) are resolved on facter3/puppet4 with the dictionary lookup on what `json.loads()` returns...